src/classes/PMT_Payment_TDTM.cls

```diff
@@ -150,25 +150,39 @@ public class PMT_Payment_TDTM extends TDTM_Runnable {
 
     /*******************************************************************************************************
     * @description Updates an opportunity's payment record when currency on the Opportunity is changed.
-    * @param listOpps A list of opportunities whose currency has changed.
+    * @param opps A list of opportunities whose currency has changed.
     * @return void This method modifies the dmlWrapper var and does not return anything
     ********************************************************************************************************/
-    private static void setPaymentCurrencyFromOpportunity(list<Opportunity> listOpps, DmlWrapper dmlWrapper){
+    private static void setPaymentCurrencyFromOpportunity(List<Opportunity> opps, DmlWrapper dmlWrapper){
+        final String currencyIsoCodeField = UTIL_Currency.CURRENCY_ISO_CODE_FIELD;
+        final String paymentAmountField = 'npe01__Payment_Amount__c';
 
-        // Get a list of the Opportunity Id's
-        Set<Id> oppIds = (new Map<Id, Opportunity>(listOpps)).keySet();
+        Set<Id> oppIds = (new Map<Id, Opportunity>(opps)).keySet();
+
+        Map<Id, SObject> dmlWrapperSObjectById = dmlWrapper.objectsToUpdate == null || dmlWrapper.objectsToUpdate.isEmpty()
+            ? new Map<Id, SObject>()
+            : new Map<Id, SObject>(dmlWrapper.objectsToUpdate);
 
         // Get a list of the Payment records by Opportunity and update the CurrencyIsoCode value
         String soql =
             'SELECT Id, CurrencyIsoCode, npe01__Payment_Amount__c, npe01__Opportunity__r.CurrencyIsoCode ' +
             'FROM npe01__OppPayment__c ' +
             'WHERE npe01__Opportunity__c IN :oppIds ' +
             'ORDER BY npe01__Opportunity__c';
+
         for (npe01__OppPayment__c pmt : database.query(soql)) {
             Opportunity opp = (Opportunity)pmt.getSObject('npe01__Opportunity__r');
-            pmt.put('CurrencyIsoCode', opp.get('CurrencyIsoCode'));
-            pmt.put('npe01__Payment_Amount__c', pmt.npe01__Payment_Amount__c);
-            dmlWrapper.objectsToUpdate.add(pmt);
+
+            if (dmlWrapperSObjectById.containsKey(pmt.Id)) {
```

Comment:
When both currency and Amount are changed on RD, then "duplicate id in list" error is generated in the `PMT_Payment_TDTM` trigger handler. Thus, this change is verifying if `dmlWrapper` has the payment already. The original change for this part of code was done in PR #2710 (see [comment](https://github.com/SalesforceFoundation/NPSP/pull/2710#discussion_r135633410)).

If the payment is already in the `dmlWrapper`, then the payment currency is updated only. If the Amount is updated from the payment as it is currently in DB, then the Amount is wrong (`RD2_OpportunityService_TEST.shouldAllowCurrencyChangeWhenRDHasNoClosedOpp()` test would fail).

Reply:
@Julian88Tex , @ravensisco  ^

Reply:
FYI @jbrandolinisf @lparrott